src/dag_combine.rs

```diff
@@ -0,0 +1,56 @@
+use digraph::PyDiGraph;
+use pyo3::prelude::*;
+use petgraph::algo;
+use pyo3::Python;
+use pyo3::types::PyAny;
+use std::cmp::Ordering;
+
+fn contains_weight(py: Python, dag: &PyDiGraph, obj: &PyObject) -> i32 {
+    let mut index :i32 = -1;
+
+    for node in dag.graph.node_indices() {
+        let weight = dag.graph.node_weight(node).unwrap();
+        if ! weight.is_none(py) {
+
+            let predicate_callable = |a: &PyAny, b: &PyAny| -> PyResult<bool> {
+                let res = a.compare(b)?;
+                Ok(res == Ordering::Equal)
+            };
+
+            if predicate_callable(obj.as_ref(py), weight.as_ref(py)).unwrap() {
+                index = node.index() as i32;
+                break;
+            }
+        }
+    }
+    return index;
+}
+
+pub fn combine_graphs(
+    py: Python,
+    dag0: &PyDiGraph,
+    dag1: &PyDiGraph,
+) -> PyResult<PyDiGraph> {
+    let first = &dag0.graph;
+    let second = &dag1.graph;
+    let mut combined = PyDiGraph {
+        graph: first.clone(),
+        cycle_state: algo::DfsSpace::default(),
+        check_cycle: false,
+        node_removed: false,
+    };
+    for node in second.node_indices(){
+        let weight = &second[node].clone();
```

Comment:
This applies to everything else in the function, but do you want `clone()` or `clone_ref(py)` to be used? Just like I asked in `merge_node` one is a ref counter bump the other copies the memory. So it depends if you want shared references or fresh copies. I think you should probably also benchmark both to see what the performance difference is.